### PR TITLE
[quant][graphmode] Fix add_relu patterns for scripting and tracing

### PR DESCRIPTION
**Pull request by jerryzh168**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #39615 [quant][graphmode] Fix a corner case in handling `if` in insert_observers
* #39551 [quant][graphmode] Support propagate dequantize for nodes with multiple outputs
* #39550 [quant][graphmode][refactor] propagateQuantizationOps
* #39490 [quant][graphmode] Support conv*d_relu in traced models
* **#39455 [quant][graphmode] Fix add_relu patterns for scripting and tracing**

Summary:
1. enable filters in PatternInfo
2. add aten_add_alpha_is_one filter
3. add is_functional_relu filter
4. add is_relu_module filter
5. fix the relu module method call matching in traced modules with regex
6. add aten::add - aten::relu patterns for traced modules

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21917118](https://our.internmc.facebook.com/intern/diff/D21917118)

### CODE REVIEW

File: test/quantization/test_quantize_script.py [hunk 1]
```diff
@@ -1465,9 +1580,7 @@ def forward(self, x):
                 FileCheck().check_not("quantized::add_scalar") \
                            .run(m.graph)
 
-    @unittest.skipUnless('fbgemm' in torch.backends.quantized.supported_engines,
-                         " Quantized operations require FBGEMM. FBGEMM is only optimized for CPUs"
-                         " with instruction set support avx2 or newer.")
+    @skipIfNoFBGEMM
```
Comment (raghuramank100) [inline test/quantization/test_quantize_script.py:1583]: Should we be able to run this test for qnnpack if fbgemm is not present?

File: torch/csrc/jit/passes/quantization/insert_observers.cpp [hunk 2]
```diff
@@ -130,7 +148,7 @@ class ModuleCloneHelper {
         r.register_attribute(
             type->getAttributeName(i),
             type->getAttribute(i),
-            inplace ? s : s.deepcopy(memo),
+            inplace || s.isCapsule() ? s : s.deepcopy(memo),
```
Comment (raghuramank100) [inline torch/csrc/jit/passes/quantization/insert_observers.cpp:151]: Should this change be part of a separate PR?

File: test/quantization/test_quantize_script.py (hunk 1 above)
Comment (jerryzh168) [reply test/quantization/test_quantize_script.py:1583]: maybe this will work, but we can enable it in a separate diff

File: torch/quantization/quantize_script.py [hunk 3]
```diff
@@ -29,6 +29,7 @@ def _prepare_script(model, qconfig_dict, inplace=False, is_dynamic=False):
     scripted_qconfig_dict = script_qconfig_dict(qconfig_dict)
     torch._C._jit_pass_dedup_module_uses(model._c)
     model = wrap_cpp_module(torch._C._jit_pass_fold_convbn(model._c))
+    print('after dedup use and fold bn:', model.graph)
```
Comment (raghuramank100) [inline torch/quantization/quantize_script.py:32]: Remove print

File: test/quantization/test_quantize_script.py [hunk 4]
```diff
@@ -1347,11 +1351,122 @@ def forward(self, x, y):
...
         m = QuantizedAdd()
         m = self._test_op_impl(m, data, "quantized::add")
-        FileCheck().check_count("quantized::add", 3, exactly=True)
+        FileCheck().check_count("quantized::add", 3, exactly=True) \
+                   .run(m.graph)
         FileCheck().check_not("aten::add") \
                    .check_not("aten::add_") \
                    .run(m.graph)
 
+    @skipIfNoFBGEMM
```
Comment (raghuramank100) [inline test/quantization/test_quantize_script.py:1360]: Why do we need to skip here? Should we just run over supported qengines?

Comment (jerryzh168) [reply test/quantization/test_quantize_script.py:1360]: I can change this in a separate diff, fbgemm definitely works so I put it here.

File: torch/csrc/jit/passes/quantization/insert_observers.cpp (hunk 2 above)
Comment (jerryzh168) [reply torch/csrc/jit/passes/quantization/insert_observers.cpp]: this change is not needed, removed now

File: test/quantization/test_quantize_script.py [hunk 5]
```diff
@@ -1347,11 +1351,122 @@ def forward(self, x, y):
...
+                y = self.conv2(y)
+                x += y
+                x = F.relu(x, True)
+                x += y
+                return F.relu(x, True)
+
+        data = [(torch.rand((1, 3, 10, 10), dtype=torch.float),
+                 torch.rand((1, 3, 10, 10), dtype=torch.float),
+                 torch.randint(0, 1, (1,), dtype=torch.long)) for _ in range(2)]
+        for m_orig in [AddRelu(True), AddRelu(False),
```
Comment (raghuramank100) [inline test/quantization/test_quantize_script.py:1454]: Is there a way to make the model set smaller? As we have more operations similar to this (sub/mul etc), it feel clunky to have a lot of models? Would a single model performing all the operation types here be better?

File: test/quantization/test_quantize_script.py (hunk 1 above)
Comment (raghuramank100) [reply test/quantization/test_quantize_script.py:1583]: Sounds good

File: test/quantization/test_quantize_script.py (hunk 5 above)
Comment (jerryzh168) [reply test/quantization/test_quantize_script.py:1454]: we can try to see if we can refactor this test with some python features, but I feel it's likely not supported in JIT.

regarding single model vs multiple model: I feel it's better to keep them separate since it's easier to see what is failing.

File: torch/csrc/jit/passes/quantization/helper.h [hunk 6]
```diff
@@ -89,7 +95,37 @@ findChildModule(const Module& module, const std::vector<std::string>& path);
...
+
+auto is_functional_relu =
+    [](const Match& match,
+       const std::unordered_map<std::string, Value*>& vmap) {
+      const auto& match_vmap = match.values_map;
+      Value* relu = match_vmap.at(vmap.at("relu"));
+      return relu->type()->cast<FunctionType>() && getFuncName(relu) == "relu";
+    };
+
+auto is_relu_module = [](const Match& match,
```
Comment (raghuramank100) [inline torch/csrc/jit/passes/quantization/helper.h:116]: Can we make these functions general, so that they take a generic function name, for example:
``` c++
auto is_functional =
    [](const Match& match,
       const std::unordered_map<std::string, Value*>& vmap,  std::string functional_name) {
      const auto& match_vmap = match.values_map;
      Value* relu = match_vmap.at(vmap.at(functional_name));
      return relu->type()->cast<FunctionType>() && getFuncName(relu) == functional_name;
    };
```

File: torch/csrc/jit/passes/quantization/helper.h [hunk 7]
```diff
@@ -89,7 +95,37 @@ findChildModule(const Module& module, const std::vector<std::string>& path);
 // Given an CallMethod node, get the module instance corresponding
 // to the instance Value
 TORCH_API Module getInvokedModule(Module& module, Node* n, Value* self);
-// =========== helper functions for Module  ==========
+
+// ==================== filter functions for matches ==============
+auto aten_add_alpha_is_one =
```
Comment (raghuramank100) [inline torch/csrc/jit/passes/quantization/helper.h:100]: How about a general matcher that checks if a particular argument is hardcoded to an integer value?
```c++
    auto check_constant_argument =
[](const Match& match,
       const std::unordered_map<std::string, Value*>& vmap, std::string argument_name, int argument_value) {
      const auto& match_vmap = match.values_map;
      auto alpha = toIValue(match_vmap.at(vmap.at(argument_name)));
      return alpha && alpha->isInt() && alpha->toInt() == argument_value;
    };

File: torch/csrc/jit/passes/quantization/insert_observers.cpp [hunk 8]
```diff
@@ -457,36 +459,60 @@ graph(%input, %weight, %bias, %4):
...
-graph(%self, %a, %b):
-     %one = prim::Constant[value=1]()
-     %first_output = aten::add(%a, %b, %one)
-     %second_module = match::module[name="ReLU"](%self)
-     %second_output = prim::CallMethod[name="forward"](%second_module, %first_output)
+  const PatternInfo add_nn_relu = PatternInfo::parse_from_str(
+      R"(
+graph(%self, %a, %b, %alpha, %relu):
+     %first_output = aten::add(%a, %b, %alpha)
+     %second_output = prim::CallMethod[name="forward.*"](%relu, %first_output)
```
Comment (raghuramank100) [inline torch/csrc/jit/passes/quantization/insert_observers.cpp:466]: Is there a convention that the callmethod always has the name of the form forward.*?

File: torch/csrc/jit/passes/quantization/insert_observers.cpp [hunk 9]
```diff
@@ -457,36 +459,60 @@ graph(%input, %weight, %bias, %4):
...
+     %second_output = aten::relu(%first_output)
      return (%second_output) )");
 
-  const PatternInfo inplace_add_f_relu = PatternInfo::parse_from_str(R"(
-graph(%self, %a, %b, %inplace):
-     %one = prim::Constant[value=1]()
-     %first_output = aten::add_(%a, %b, %one)
-     %relu = prim::Constant[name="relu"]()
-     %second_output = prim::CallFunction(%relu, %first_output, %inplace)
+  const PatternInfo inplace_add_aten_relu_ = PatternInfo::parse_from_str(R"(
```
Comment (raghuramank100) [inline torch/csrc/jit/passes/quantization/insert_observers.cpp:512]: Can we generate these patterns to make the code readable? If you iterate over the patterns (add, add_) and (relu, relu_) can you simplify this part?

**Review by raghuramank100 — CHANGES_REQUESTED** [review]
Please see comments

File: torch/csrc/jit/passes/quantization/insert_observers.cpp (hunk 8 above)
Comment (jerryzh168) [reply torch/csrc/jit/passes/quantization/insert_observers.cpp:466]: this is to handle methods produced by tracing, I'll probably use forward followed by digits to be more accurate

Comment (jerryzh168) [reply torch/csrc/jit/passes/quantization/insert_observers.cpp:466]: callmethod can have any name.

File: torch/csrc/jit/passes/quantization/insert_observers.cpp (hunk 9 above)
Comment (jerryzh168) [reply torch/csrc/jit/passes/quantization/insert_observers.cpp:512]: this is the third starter task for my intern, so let's land this asap

**Review by raghuramank100 — APPROVED** [review]
